# Pandas passo a passo — CSV — padrões de texto com regex

Arquivo: **pedidos_regex.csv**. Dados fictícios de clientes e pedidos. valor está em reais.

## Abrir o arquivo

No Colab: na barra esquerda, clique na pasta Arquivos e no botão de upload. Envie o arquivo de dados indicado abaixo. No VS Code: coloque o arquivo na mesma pasta do notebook e abra essa pasta como projeto.

Se surgir `FileNotFoundError`, confira nome e pasta. No Colab, os arquivos enviados podem desaparecer ao encerrar a sessão; envie novamente quando necessário.

Não execute todos os notebooks juntos: cada atividade usa sua própria variável `df`.

**Conceito novo:** regex, ou expressão regular, descreve um padrão de texto. Vamos verificar códigos no formato PED- seguido de exatamente três dígitos e retirar pontuação dos telefones. Esta verificação é de formato; não comprova que um pedido ou telefone existe.

| Símbolo | Significado |
|---|---|
| `PED-` | Texto literal |
| `[0-9]` | Um dígito de 0 a 9 |
| `{3}` | Exatamente três repetições |
| `\D` | Um caractere que não é dígito |

`r"..."` é uma string bruta, útil para escrever barras invertidas nos padrões. `str.fullmatch` verifica o texto inteiro; `str.replace` substitui partes que combinam com o padrão.

## Como usar


1. Na primeira parte, tente resolver sem olhar o gabarito.
2. Escreva o código nas células de código e suas explicações nas células de texto.
3. Execute as células de cima para baixo.

**Como pensar:** qual informação preciso? Qual coluna contém essa informação? Quero todas as linhas ou apenas algumas? Qual operação responde à pergunta?

Comentários iniciados por `#` não são executados. `NameError` pode indicar uma preparação não executada; `KeyError` pode indicar um nome de coluna escrito incorretamente.

## Preparação

`import pandas as pd` disponibiliza a biblioteca com o apelido `pd`. A função de leitura abre o arquivo e cria um DataFrame. `df =` guarda a tabela na variável `df`. O arquivo precisa estar na pasta de trabalho; o notebook não cria o arquivo de entrada.

Os números à esquerda são os índices das linhas. Eles começam em zero.

In [41]:
import pandas as pd

df = pd.read_csv("pedidos_regex.csv", dtype={"telefone": "string"}, encoding="utf-8")
df

,cliente,codigo,telefone,valor
0,Ana,PED-001,(41) 90000-0001,40
1,Bruno,PED-002,41 90000 0002,120
2,Carla,XYZ-003,41-90000-0003,150
3,Diego,PED-004,(11) 90000-0004,200
4,Elisa,PED-ABC,11 90000 0005,60
5,Felipe,PED-006,(41) 90000-0006,90


## Parte 1 — perguntas

Para cada questão: **preveja → escreva → execute → interprete**.

### Questão 1

Mostre as primeiras linhas e o tamanho da tabela.

**Pense antes:** O que cada linha representa?

**Minha previsão:**

Escreva aqui.

In [42]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 2

Inspecione as colunas e os tipos.

**Pense antes:** Telefone deve ser número para somar?

**Minha previsão:**

Escreva aqui.

In [43]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 3

Mostre nome e código usando loc.

**Pense antes:** Quais códigos parecem diferentes?

**Minha previsão:**

Escreva aqui.

In [44]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 4

Calcule a média dos valores dos pedidos.

**Pense antes:** Qual coluna usar?

**Minha previsão:**

Escreva aqui.

In [45]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 5

Verifique quais códigos seguem PED- e três dígitos.

**Pense antes:** O código PED-ABC atende? E XYZ-003?

**Minha previsão:**

Escreva aqui.

In [46]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 6

Mostre os pedidos com códigos fora do padrão.

**Pense antes:** Como inverter a seleção booleana?

**Minha previsão:**

Escreva aqui.

In [47]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 7

Crie uma cópia e uma coluna telefone_limpo só com dígitos.

**Pense antes:** O que deve sair: espaços, parênteses e hífens?

**Minha previsão:**

Escreva aqui.

In [48]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 8

Mostre os clientes cujo telefone limpo começa com 41.

**Pense antes:** Qual operação simples verifica o início de um texto?

**Minha previsão:**

Escreva aqui.

In [49]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

### Questão 9

Exporte os dados com a nova coluna para CSV.

**Pense antes:** Vamos sobrescrever o arquivo de entrada?

**Minha previsão:**

Escreva aqui.

In [50]:
# Escreva sua solução abaixo.


**Minha interpretação do resultado:**

Escreva aqui.

## Parte 2 — resolução comentada

Compare com suas tentativas. Leia como pensar, execute o código e explique o resultado com suas palavras. Comece executando a preparação novamente.

In [51]:
import pandas as pd

df = pd.read_csv("pedidos_regex.csv", dtype={"telefone": "string"}, encoding="utf-8")
df

,cliente,codigo,telefone,valor
0,Ana,PED-001,(41) 90000-0001,40
1,Bruno,PED-002,41 90000 0002,120
2,Carla,XYZ-003,41-90000-0003,150
3,Diego,PED-004,(11) 90000-0004,200
4,Elisa,PED-ABC,11 90000 0005,60
5,Felipe,PED-006,(41) 90000-0006,90


### Resolução 1

**Pedido:** Mostre as primeiras linhas e o tamanho da tabela.

**Como pensar:** Começamos olhando os dados antes de escrever o padrão.

In [52]:
print(df.shape)
df.head()

(6, 4)


,cliente,codigo,telefone,valor
0,Ana,PED-001,(41) 90000-0001,40
1,Bruno,PED-002,41 90000 0002,120
2,Carla,XYZ-003,41-90000-0003,150
3,Diego,PED-004,(11) 90000-0004,200
4,Elisa,PED-ABC,11 90000 0005,60


**Resultado explicado:** 6 pedidos e 4 colunas.

### Resolução 2

**Pedido:** Inspecione as colunas e os tipos.

**Como pensar:** Telefone é informação textual, não uma quantidade. Na leitura pedimos dtype string para essa coluna.

In [61]:
print(df.columns)
df.info()

Index(['cliente', 'codigo', 'telefone', 'valor'], dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6 entries, 0 to 5
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   cliente   6 non-null      object
 1   codigo    6 non-null      object
 2   telefone  6 non-null      string
 3   valor     6 non-null      int64 
dtypes: int64(1), object(2), string(1)
memory usage: 320.0+ bytes


**Resultado explicado:** cliente, codigo, telefone e valor. Não devemos calcular média de telefone.

### Resolução 3

**Pedido:** Mostre nome e código usando loc.

**Como pensar:** Selecionamos duas colunas para examinar os exemplos antes da regex.

In [54]:
df.loc[:, ["cliente", "codigo"]]

,cliente,codigo
0,Ana,PED-001
1,Bruno,PED-002
2,Carla,XYZ-003
3,Diego,PED-004
4,Elisa,PED-ABC
5,Felipe,PED-006


**Resultado explicado:** XYZ-003 e PED-ABC fogem do formato pretendido.

### Resolução 4

**Pedido:** Calcule a média dos valores dos pedidos.

**Como pensar:** Selecionamos a coluna numérica e calculamos mean.

In [55]:
df["valor"].mean()

np.float64(110.0)

**Resultado explicado:** A soma é 660; a média é 110 reais.

### Resolução 5

**Pedido:** Verifique quais códigos seguem PED- e três dígitos.

**Como pensar:** PED- é fixo; [0-9] é um dígito; {3} exige três. fullmatch exige correspondência do texto inteiro. na=False trata uma eventual ausência como não correspondente.

In [56]:
padrao = r"PED-[0-9]{3}"
codigo_valido = df["codigo"].str.fullmatch(padrao, na=False)
print(codigo_valido)

0     True
1     True
2    False
3     True
4    False
5     True
Name: codigo, dtype: bool


**Resultado explicado:** True para Ana, Bruno, Diego e Felipe. False para Carla e Elisa.

### Resolução 6

**Pedido:** Mostre os pedidos com códigos fora do padrão.

**Como pensar:** O símbolo ~ inverte True e False nessa Series booleana.

In [57]:
df.loc[~codigo_valido, ["cliente", "codigo"]]

,cliente,codigo
2,Carla,XYZ-003
4,Elisa,PED-ABC


**Resultado explicado:** Carla: XYZ-003; Elisa: PED-ABC.

### Resolução 7

**Pedido:** Crie uma cópia e uma coluna telefone_limpo só com dígitos.

**Como pensar:** \D encontra caracteres que não são dígitos. Substituímos cada ocorrência por texto vazio. regex=True ativa a interpretação do padrão. Preservamos a coluna original.

In [58]:
df_limpo = df.copy()
df_limpo["telefone_limpo"] = df_limpo["telefone"].str.replace(r"\D", "", regex=True)
df_limpo.loc[:, ["telefone", "telefone_limpo"]]

,telefone,telefone_limpo
0,(41) 90000-0001,41900000001
1,41 90000 0002,41900000002
2,41-90000-0003,41900000003
3,(11) 90000-0004,11900000004
4,11 90000 0005,11900000005
5,(41) 90000-0006,41900000006


**Resultado explicado:** Ana fica com 41900000001; Bruno com 41900000002. Limpar a pontuação não verifica se o telefone é real.

### Resolução 8

**Pedido:** Mostre os clientes cujo telefone limpo começa com 41.

**Como pensar:** Depois da limpeza, startswith verifica o prefixo. Para um prefixo literal simples não é necessário usar regex.

In [59]:
df_limpo.loc[df_limpo["telefone_limpo"].str.startswith("41", na=False), "cliente"]

0       Ana
1     Bruno
2     Carla
5    Felipe
Name: cliente, dtype: object

**Resultado explicado:** Ana, Bruno, Carla e Felipe.

### Resolução 9

**Pedido:** Exporte os dados com a nova coluna para CSV.

**Como pensar:** Usamos outro nome para criar uma saída com os textos originais e a coluna limpa.

In [60]:
df_limpo.to_csv("pedidos_limpos.csv", index=False, encoding="utf-8")

**Resultado explicado:** pedidos_limpos.csv é criado. No Colab, use o menu do arquivo para baixar.

## Lembretes

| Recurso | Significado |
|---|---|
| `df["coluna"]` | Seleciona uma coluna |
| `df[condicao]` | Seleciona as linhas que atendem à condição |
| `df.loc[linhas, colunas]` | Seleciona por rótulos ou condições |
| `df.iloc[linhas, colunas]` | Seleciona por posições começando em zero |
| `:` sozinho | Todas as linhas ou colunas |
| `0:3` no iloc | Posições 0, 1 e 2; exclui a posição 3 |
| `==` | Compara valores |
| `=` | Atribui um valor |

Selecionar ou mostrar um resultado não significa apagar as outras informações. A alteração de valores é feita explicitamente, como na questão 9.

Documentação: https://pandas.pydata.org/docs/getting_started/intro_tutorials/index.html